# Exercises 1

In [225]:
import warnings 
warnings.filterwarnings("ignore")

In [7]:
#Scenario: Build a classifier to predict whether a given face should unlock the iPhone.

#What is the positive and negative case?
#postive: face should unlock the iphone. 
#negative: face shouldn't unlock the iphone. 

#What are the possible outcomes?
#The phone's owner ---->unclocked the phone. 
#The phone's owner ----> can't unclock the phone. 
# Somebody else ----->unclocked the phone. 
# somebody else -----> can't unlock the phone. 


#What are the costs of the outcomes? The cost of FP is higher than FN. 
# if anyone can unlock the phone when the cost of fix the phone will occur. 
# privacy is not protected. 

#Which metric should we use?
#precision.



In [8]:
#Scenario: Predict whether an email is spam or not. Emails marked as spam skip the inbox and go to the spam folder.

#What is the positive and negative case?
#mark an email is spam
#mark an email is not a spam.

#What are the possible outcomes?
#email is spam ---> spam folder
#email is spam ---> main folder 
#email is not spam ---> spam folder
#email is not spam ---> main folder

#What are the costs of the outcomes?
# The cost of being missing an important email but went to spam. 


#Which metric should we use?
# precision, because missing an important email is more worse. 


In [9]:
# Scenario: Predict whether an email is a phishing attempt.
#When we predict positive, show an additional banner warning the user 
#that this might be a phishing email.

#What is the positive and negative case?
#What are the possible outcomes?
#What are the costs of the outcomes?
#Which metric should we use?
#recall 

## Exercises 2

## Given the following confusion matrix, evaluate (by hand) the model's performance.

In [38]:
df = pd.DataFrame({
    'pred dog': [46, 13],
    'pred cat': [7, 34]},
    index=['actual dog', 'actual cat']
)
df

pred dog  pred cat
actual dog        46         7
actual cat        13        34

In [15]:
# dog = positive class
# cat = negative class

### In the context of this problem, what is a false positive?

FN: Predicted dog ---> Actual is not a dog, a cat

### In the context of this problem, what is a false negative?

FN: Predict is not a dog, a cat ----> Actual is a dog. 

### How would you describe this model?

TP: predicted dog + actual is dog   46
    
    
FP: predicted dog, but is not dog.  7


FN: predicted not dog, but is dog.  13


TN: predicted not dog, and is not dog.   34

In [23]:
tp = 46 
fp = 7
fn = 13
tn = 34

In [24]:
recall = tp/(tp+fn)
recall 

0.7796610169491526

In [25]:
precision = tp / (tp + fp)
precision

0.8679245283018868

In [258]:
accuracy = (tp + tn) / (tp + tn + fp + fn)
accuracy

0.6508

In [257]:
df = pd.DataFrame({
    'pred dog': [53, 47],
    'pred cat': [0, 0]},
    index=['actual dog', 'actual cat']
)
df

pred dog  pred cat
actual dog        53         0
actual cat        47         0

In [48]:
tp = 53
tn = 0
fp = 47
fn = 0


accuracy = (tp + tn) / (tp + tn + fp + fn)
accuracy

0.53

In [49]:
### The model is better than our baseline model so this model is helpful. 

## Exercises 3

### You are working as a datascientist working for Codeup Cody Creator (C3 for short), a rubber-duck manufacturing plant.

Unfortunately, some of the rubber ducks that are produced will have defects. Your team has built several models that try to predict those defects, and the data from their predictions can be found here.

Use the predictions dataset and pandas to help answer the following questions:

In [283]:
df = pd.read_csv('c3.csv')

In [284]:
df.head(2)

actual     model1  model2     model3
0  No Defect  No Defect  Defect  No Defect
1  No Defect  No Defect  Defect     Defect

### An internal team wants to investigate the cause of the manufacturing defects. They tell you that they want to identify as many of the ducks that have a defect as possible. Which evaluation metric would be appropriate here? Which model would be the best fit for this use case?

In [67]:
# recall -- how often do we get the actual positive cases
# recall -- model performance | actual +

In [295]:
columns = ['model1','model2','model3']
for column in columns: 
    subset = df[df.actual== 'Defect']
    rate =(subset.actual == subset[column]).mean()
    print(column)
    print(f'The recall reate is {rate:.2%}')

model1
The recall reate is 50.00%
model2
The recall reate is 56.25%
model3
The recall reate is 81.25%


In [ ]:
TP: predicted defect + actual is defect
FP: predicted defect , actual not defect
FN: predicted no defect, but is defect 
TN: predicted no defect, actual no defect

In [88]:
recalls = []
columns = ['model1','model2','model3']

for column in columns: 
    postive = df[column] == 'Defect'
    correct = df.actual == df[column]
    tp = df[postive & correct].shape[0] # to get the number of row shape 
    
    wrong = df.actual != df[column]
    negative = df[column] != "Defect"
    fn = df[wrong & negative].shape[0]

    
    recalls.append(tp/(tp+fn))
recalls

[0.5, 0.5625, 0.8125]

In [296]:
# model 3 best fit for recall matrix. 81.25%

### Recently several stories in the local news have come out highlighting customers who received a rubber duck with a defect, and portraying C3 in a bad light. The PR team has decided to launch a program that gives customers with a defective duck a vacation to Hawaii. They need you to predict which ducks will have defects, but tell you the really don't want to accidentally give out a vacation package when the duck really doesn't have a defect. Which evaluation metric would be appropriate here? Which model would be the best fit for this use case?

Precision

In [288]:
for column in columns: 
    print(column)
    subset = df[df[column]=="Defect"]
    model = (subset.actual ==subset[column]).mean()
    print(f'model precision: {model:.2%}')
    print("-------------")


model1
model precision: 80.00%
-------------
model2
model precision: 10.00%
-------------
model3
model precision: 13.13%
-------------


In [77]:
# model 1 best fit for precision matrix. 

## Exercises 4

### You are working as a data scientist for Gives You Paws ™, a subscription based service that shows you cute pictures of dogs or cats (or both for an additional fee).

At Gives You Paws, anyone can upload pictures of their cats or dogs. The photos are then put through a two step process. First an automated algorithm tags pictures as either a cat or a dog (Phase I). Next, the photos that have been initially identified are put through another round of review, possibly with some human oversight, before being presented to the users (Phase II).

Several models have already been developed with the data, and you can find their results here.

In [298]:
df = pd.read_csv('gives_you_paws.csv')

In [299]:
df.head(2)

actual model1 model2 model3 model4
0    cat    cat    dog    cat    dog
1    dog    dog    cat    cat    dog

In [300]:
df['baseline'] ='dog'
df['baselinecat']='cat'

In [263]:
df.head(2)

actual model1 model2 model3 model4 baseline baselinecat
0    cat    cat    dog    cat    dog      dog         cat
1    dog    dog    cat    cat    dog      dog         cat

### Model 1 - 4 and baseline accuracy. 

In [274]:
accuracy = []
columns = ['model1','model2','model3','model4','baseline','baselinecat']

for column in columns: 
    value = (df.actual == df[column]).mean()
    accuracy.append(value)
    print(column)
    print(f'model accuracy: {value:.2%}')
    print("-------------------")


model1
model accuracy: 80.74%
-------------------
model2
model accuracy: 63.04%
-------------------
model3
model accuracy: 50.96%
-------------------
model4
model accuracy: 74.26%
-------------------
baseline
model accuracy: 65.08%
-------------------
baselinecat
model accuracy: 34.92%
-------------------


 The model 1 have the highest accuracy. 80%

### Phase 1: model 1 - 4 and recall rate.

In [172]:
wrong = df[df.actual != df.model1]
negative = df[df.model1 != 'dog']
negative.head(2)
#recall = TP/TP+NF
#Wrong求的是actual ！= predict
#negative 求的是 predict ！= dog
# NF 求的是，predict 没有猜狗，但是实际是狗，像下面的2 就会被算进去，但0 不会因为没有猜狗，也不是狗，猜对了。

actual model1 model2 model3 model4 baseline
0    cat    cat    dog    cat    dog      dog
2    dog    cat    cat    cat    dog      dog

In [185]:
a = df[(df.actual != df.model1) & (df.model1 != 'dog')].shape[0]

In [186]:
b = df[(df.model1 == 'dog') & (df.actual == df.model1)].shape[0]

In [190]:
b/(a+b)

0.803318992009834

In [134]:
recalls = []
columns = ['model1','model2','model3','model4','baseline']

for column in columns: 
    postive = df[column] == 'dog'
    correct = df.actual == df[column]
    tp = df[postive & correct].shape[0] # to get the number of row shape 
    
    wrong = df.actual != df[column]
    negative = df[column] != "dog"
    fn = df[wrong & negative].shape[0]  # here guess cat but actual is dof
    recalls.append(tp/(tp+fn))
recalls

[0.803318992009834,
 0.49078057775046097,
 0.5086047940995697,
 0.9557467732022127,
 1.0]

In [301]:
columns = ['model1','model2','model3','model4','baseline']
for column in columns: 
    subset = df[df.actual== 'dog']
    rate =(subset.actual == subset[column]).mean()
    print(column)
    print(f'The recall reate is {rate:.2%}')

model1
The recall reate is 80.33%
model2
The recall reate is 49.08%
model3
The recall reate is 50.86%
model4
The recall reate is 95.57%
baseline
The recall reate is 100.00%


The model 4 has highest recall rate. 95.57%

###  Phase 2: model 1 - 4 and precision rate. 

In [275]:
for column in columns: 
    print(column)
    subset = df[df[column]=="dog"]
    model = (subset.actual ==subset[column]).mean()
    print(f'model precision: {model:.2%}')# :2 is save two decimal, % add % after round 2. 
    #it's better than round, becuase round is only change the display, but :2 and %, change the number. 
    print("-------------")

model1
model precision: 89.00%
-------------
model2
model precision: 89.32%
-------------
model3
model precision: 65.99%
-------------
model4
model precision: 73.12%
-------------
baseline
model precision: 65.08%
-------------
baselinecat
model precision: nan%
-------------


In [246]:
wrong = df[df.actual != df.model1] 
#Pression: TP/ TP+ EP 
#wrong 求的是predict 的东西和 actual 不一样
#postive 求的是predict的是狗，管他predict的对不对，predict 的是狗就算。
# EP = 求的是 predict 是狗，但是actual 又不是狗。 7 will be counted 


#postive = df[column] == 'dog'
#correct = df.actual == df[column]
#postive 求的是predict  == dog
#correct 求的是predict == turth 
#TP = predict == dog == turth
#tp = df[postive & correct].shape[0]

wrong.head(2)

actual model1 model2 model3 model4 baseline baselinecat
2    dog    cat    cat    cat    dog      dog         cat
7    cat    dog    cat    cat    dog      dog         cat

In [248]:
pression = []
columns = ['model1','model2','model3','model4','baseline']

for column in columns: 
    postive = df[column] == 'dog'
    correct = df.actual == df[column]
    tp = df[postive & correct].shape[0] # to get the number of row shape 
    
    wrong = df.actual != df[column]
    fp = df[postive & wrong].shape[0]
    pression.append(tp/(tp+fp))
pression

[0.8900238338440586,
 0.8931767337807607,
 0.6598883572567783,
 0.7312485304490948,
 0.6508]

Model 2 has highest pression rate, 89.32%

### Suppose you are working on a team that solely deals with cat pictures. Which of these models would you recomend for Phase I? For Phase II?

In [191]:
# cat = postive class
# dog = negative class

In [193]:
df.head(2)

actual model1 model2 model3 model4 baseline baselinecat
0    cat    cat    dog    cat    dog      dog         cat
1    dog    dog    cat    cat    dog      dog         cat

In [305]:
columns = ['model1','model2','model3','model4','baseline','baselinecat']
for column in columns: 
    subset = df[df.actual== 'cat']
    rate =(subset.actual == subset[column]).mean()
    print(column)
    print(f'The recall rate is {rate:.2%}')

model1
The recall rate is 81.50%
model2
The recall rate is 89.06%
model3
The recall rate is 51.15%
model4
The recall rate is 34.54%
baseline
The recall rate is 0.00%
baselinecat
The recall rate is 100.00%


In [306]:
for column in columns: 
    print(column)
    subset = df[df[column]=="cat"]
    model = (subset.actual ==subset[column]).mean()
    print(f'model precision: {model:.2%}')# :2 is save two decimal, % add % after round 2. 
    #it's better than round, becuase round is only change the display, but :2 and %, change the number. 
    print("-------------")

model1
model precision: 68.98%
-------------
model2
model precision: 48.41%
-------------
model3
model precision: 35.83%
-------------
model4
model precision: 80.72%
-------------
baseline
model precision: nan%
-------------
baselinecat
model precision: 34.92%
-------------


If we only work on cat ( as the postieve) We will use recall as phase I because we don't want to miss out any cat, the best performance is model 2. 

The phase II we will use precision, the best performance model is also model 4. 

## Follow the links below to read the documentation about each function, then apply those functions to the data from the previous problem.

sklearn.metrics.accuracy_score

sklearn.metrics.precision_score

sklearn.metrics.recall_score

sklearn.metrics.classification_report

## Accuracy

In [204]:
from sklearn.metrics import accuracy_score

In [212]:
[accuracy_score(df.actual,df[column]) for column in columns]

[0.8074, 0.6304, 0.5096, 0.7426, 0.3492]

In [ ]:
subset = df[df.actual== 'dog']
[(subset.actual == subset[column]).mean() for column in columns]

## Precision

In [217]:

import sklearn.metrics as skm

In [226]:
[skm.precision_score(df.actual,df[column], pos_label='dog') for column in columns]

[0.8900238338440586,
 0.8931767337807607,
 0.6598883572567783,
 0.7312485304490948,
 0.0]

## Recall

In [219]:
[skm.recall_score(df.actual,df[column], pos_label='dog') for column in columns]

[0.803318992009834,
 0.49078057775046097,
 0.5086047940995697,
 0.9557467732022127,
 0.0]

In [227]:
for column in columns: 
    print(column)
    print(skm.classification_report(df.actual,df[column]))


model1
              precision    recall  f1-score   support

         cat       0.69      0.82      0.75      1746
         dog       0.89      0.80      0.84      3254

    accuracy                           0.81      5000
   macro avg       0.79      0.81      0.80      5000
weighted avg       0.82      0.81      0.81      5000

model2
              precision    recall  f1-score   support

         cat       0.48      0.89      0.63      1746
         dog       0.89      0.49      0.63      3254

    accuracy                           0.63      5000
   macro avg       0.69      0.69      0.63      5000
weighted avg       0.75      0.63      0.63      5000

model3
              precision    recall  f1-score   support

         cat       0.36      0.51      0.42      1746
         dog       0.66      0.51      0.57      3254

    accuracy                           0.51      5000
   macro avg       0.51      0.51      0.50      5000
weighted avg       0.55      0.51      0.52      5000

40.0